arch/x86/kvm/pkvm/pgtable.c

```diff
@@ -752,3 +773,48 @@ int pkvm_pgtable_set_owner(struct pkvm_pgtable *pgt, unsigned long vaddr,
 
 	return pkvm_pgtable_walk(pgt, vaddr, size, &walker);
 }
+
+/**
+ * pkvm_pgtable_check_and_map() - Validates the owner id before maps a physical
+ *				  address range in a page table.
+ * @pgt:		Pointer to the page table structure.
+ * @vaddr:		Virtual address where the mapping should start.
+ * @phys:		Physical address to be mapped.
+ * @size:		Size of the mapping in bytes.
+ * @prot:		Property bits for the new mapping.
+ * @expt_owner_id:	The expected owner ID stored in the current mapping.
+ *
+ * The checked and mapped range is the minimum PAGE_SIZE-aligned range covering
+ * [@vaddr, @vaddr + @size), and @phys is aligned down to the PAGE_SIZE.
+ *
+ * Walks the page table to install @vaddr to @phys mappings for the specified
+ * virtual address range [@vaddr, @vaddr + @size) with @prot. Before install
+ * mapping for each leaf, check if the owner ID stored in the leaf matches with
+ * @expt_owner_id or not. The mapping will only be installed if they match.
```

Comment:
This doesn't clarify that if we've already mapped some pages (which passed the owner id check) while walking, and then the next walked leaf fails the check, we do *not* remove the previously installed mappings.

...In fact, I'm not entirely sure it's ok to leave the mappings in such an inconsistent state and just return an error and continue running, without a proper cleanup.

And it seems not trivial to implement restoring them to the original state on failure (we would need to somehow remember the original states of _all_ the leaves we've modified)? So perhaps it's better to not try to be smart, and just do what pKVM-ARM does, i.e. walk the page table twice: first to check the entire range, then to map the entire range?